### PR TITLE
APL-1074 The Dex Contracts Processor task moved under the control of …

### PR DESCRIPTION
**Pull request by alzinchenko**
The Dex Contracts Processor task moved under the control of the task dispatcher;
added suspend/resume methods into DexOrderProcessor;
small logging enhancement;

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/AplCore.java [hunk 1]
```diff
@@ -301,30 +294,21 @@ private void startUp() {
...
-//signal to API that core is reaqdy to serve requests. Should be removed as soon as all API will be on RestEasy                
+                //signal to API that core is ready to serve requests. Should be removed as soon as all API will be on RestEasy
                 ApiSplitFilter.isCoreReady = true;
 
                 PrunableArchiveMigrator migrator = CDI.current().select(PrunableArchiveMigrator.class).get();
                 migrator.migrate();
                 // start shard process recovery after initialization of all derived tables but before launching threads (blockchain downloading, transaction processing)
                 recoverSharding();
 
+                if(!dexOrderProcessor.isInitialized()){
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/AplCore.java]: I suggest do not throw exception here, but write log with warn level to notify user. DexOrderProcessor is an optional component of apl-core.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/BlockchainProcessorImpl.java [hunk 2]
```diff
@@ -480,6 +487,7 @@ public void processPeerBlock(JSONObject request) throws AplException {
     public void fullReset() {
         globalSync.writeLock();
         try {
+            lookupDexOrderProcessor().suspendContractProcessor();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/BlockchainProcessorImpl.java]: Not the best idea, try to decouple exchange and related classes from the core blockchain logic

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/UpdaterMediatorImpl.java [hunk 3]
```diff
@@ -52,13 +54,15 @@ public void shutdownApplication() {
 
     @Override
     public void suspendBlockchain() {
+        lookupDexOrderProcessor().suspendContractProcessor();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/UpdaterMediatorImpl.java]: Better to emit event to disable dexOrder processor (do not allow coupling between updater and exchange)

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/APIServlet.java [hunk 4]
```diff
@@ -222,6 +224,12 @@ private void process(HttpServletRequest req, HttpServletResponse resp) throws IO
             }
         } catch (ParameterException e) {
             response = e.getErrorResponse();
+        } catch (BlockNotFoundException e){
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/APIServlet.java]: You can collapse this block and move exception handling  to `AplException | RuntimeException e` block

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/post/PopOff.java [hunk 5]
```diff
@@ -54,7 +54,9 @@ public JSONStreamAware processRequest(HttpServletRequest req) {
         boolean keepTransactions = "true".equalsIgnoreCase(req.getParameter("keepTransactions"));
         List<? extends Block> blocks;
         lookupBlockchainProcessor();
+        lookupDexOrderProcessor();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/post/PopOff.java]: Better to suspend it by event emitted from blockchainProcessor

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/post/Scan.java [hunk 6]
```diff
@@ -52,8 +51,10 @@ public JSONStreamAware processRequest(HttpServletRequest req) {
                 height = Integer.parseInt(req.getParameter("height"));
             } catch (NumberFormatException ignore) {}
             long start = System.currentTimeMillis();
-            BlockchainProcessor blockchainProcessor = lookupBlockchainProcessor();
+            lookupBlockchainProcessor();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/post/Scan.java]: Same as above

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexOrderProcessor.java [hunk 7]
```diff
@@ -63,13 +73,17 @@
     private IDexValidator dexValidator;
     private DexSmartContractService dexSmartContractService;
     private EthereumWalletService ethereumWalletService;
+    private TaskDispatchManager taskDispatchManager;
+
+    private volatile boolean processorEnabled = true;
+    @Getter
+    private boolean initialized = false;
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexOrderProcessor.java]: Seems like this field is not using

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/BlockchainProcessorImpl.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/BlockchainProcessorImpl.java]: > Not the best idea, try to decouple exchange and related classes
Try explain idea... E.g. I don't understand. do you mean Inject DexOrderProcessor via BlockchainProcessorImpl constructor ?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/APIServlet.java (hunk 4 above)
Comment (alzinchenko) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/APIServlet.java]: I've split those exceptions in purpose. There isn't the stacktrace output in the BlockNotFound catch block.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexOrderProcessor.java (hunk 7 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexOrderProcessor.java]: I see: 
 @PostConstruct
    public void init(){.... initialized = true;

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/post/PopOff.java (hunk 5 above)
Comment (alzinchenko) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/post/PopOff.java]: done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/post/Scan.java (hunk 6 above)
Comment (alzinchenko) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/http/post/Scan.java]: done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/UpdaterMediatorImpl.java (hunk 3 above)
Comment (alzinchenko) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/UpdaterMediatorImpl.java]: done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/AplCore.java (hunk 1 above)
Comment (alzinchenko) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/app/AplCore.java]: changed to warning

**Pull request merged**